viv_utils/emulator_drivers.py

```diff
@@ -31,141 +28,178 @@ def __str__(self):
         return "InstructionRangeExceededError(ended at instruction 0x%08X)" % self.pc
 
 
-class Hook(LoggingObject):
-    def __init__(self):
-        super(Hook, self).__init__()
-
-    def hook(self, callname, emu, callconv, api, argv):
-        # must return something other than None if handled
-        # raise UnsupportedFunction to pass
-        raise UnsupportedFunction()
-
-
-class Monitor(vivisect.impemu.monitor.EmulationMonitor, LoggingObject):
-    def __init__(self, vw):
-        vivisect.impemu.monitor.EmulationMonitor.__init__(self)
-        LoggingObject.__init__(self)
-        self._vw = vw
-        self._logger = logging.getLogger("Monitor")
-
-    def getStackValue(self, emu, offset):
-        return emu.readMemoryFormat(emu.getStackCounter() + offset, "<P")[0]
-
-    def dumpStack(self, emu, num):
-        self._logger.debug("stack: ESP: %s", hex(emu.getStackCounter()))
-        for i in range(num):
-            self.d("stack: ESP + %s: %s",
-                               hex(emu.imem_psize * i),
-                               hex(self.getStackValue(emu, emu.imem_psize * i)))
-
+DataType: TypeAlias = str
+SymbolName: TypeAlias = str
+
+CallingConvention: TypeAlias = str
+ReturnType: TypeAlias = DataType
+ReturnName: TypeAlias = str
+FunctionName: TypeAlias = SymbolName
+ArgType: TypeAlias = DataType
+ArgName: TypeAlias = SymbolName
+FunctionArg: TypeAlias = Tuple[ArgType, ArgName]
+# type returned by `vw.getImpApi`
+API: TypeAlias = Tuple[ReturnType, ReturnName, CallingConvention, FunctionName, List[FunctionArg]]
+# shortcut
+Emulator: TypeAlias = vivisect.impemu.emulator.WorkspaceEmulator
+
+# a hook overrides an API encountered by an emulator.
+#
+# returning True indicates the hook handled the function.
+# this should include returning from the function and cleaning up the stack, if appropriate.
+# a hook can also raise `StopEmulation` to ...stop the emulator.
+#
+# hooks can fetch the current $PC, registers, mem, etc. via the provided emulator parameter.
+#
+# a hook is a callable, such as a function or class with `__call__`,
+# if the hook is "stateless", use a simple function:
+#
+#     hook_OutputDebugString(emu, api, argv):
+#         _, _, cconv, name, _ = api
+#         if name != "kernel32.OutputDebugString": return False
+#         logger.debug("OutputDebugString: %s", emu.readString(argv[0]))
+#         cconv.execCallReturn(emu, 0, len(argv))
+#         return True
+#
+# if the hook is "stateful", such as a hook that records arguments, use a class:
+#
+#     class CreateFileAHook:
+#         def __init__(self):
+#             self.paths = set()
+#
+#         def __call__(self, emu, api, argv):
+#             _, _, cconv, name, _ = api
+#             if name != "kernel32.CreateFileA": return False
+#             self.paths.add(emu.readString(argv[0]))
+#             cconv.execCallReturn(emu, 0, len(argv))
+#             return True
+#
+Hook = Callable[[Emulator, API, List[int]], bool]
```

Comment:
How about adding a `NopHook` helper or something? FLOSS could use that quite a bit (see e.g. https://github.com/mandiant/flare-floss/pull/499/files#diff-75983f483110f7478f670c58faccc16218afe8aae2469fe8a8c2d522e1d00e54R397-R403) but we could also add it there.